tests/speculos/ethereum_client/ethereum_cmd_builder.py

```diff
@@ -0,0 +1,416 @@
+
+        if self.debug:
+            logging.info("header: %s", header.hex())
+            logging.info("cdata:  %s", cdata.hex())
+
+        return header + cdata
+
+    def get_configuration(self) -> bytes:
+        """Command builder for GET_CONFIGURATON
+
+        Returns
+        -------
+        bytes
+            APDU command for GET_CONFIGURATON
+
+        """
+        return self.serialize(cla=self.CLA,
+                              ins=InsType.INS_GET_CONFIGURATION,
+                              p1=0x00,
+                              p2=0x00,
+                              cdata=b"")
+
+    def set_plugin(self, plugin: Plugin) -> bytes:
+        """Command builder for SET_PLUGIN.
+
+        Parameters
+        ----------
+            -> Check documentation of APDU
+
+        Returns
+        -------
+        bytes
+            APDU command for SET_PLUGIN.
+
+        """
+
+        cdata: bytes = plugin.serialize()
+
+        return self.serialize(cla=self.CLA,
+                              ins=InsType.INS_SET_PLUGIN,
+                              p1=0x00,
+                              p2=0x00,
+                              cdata=cdata)
+
+    def provide_nft_information(self, plugin: Plugin) -> bytes:
+        """Command builder for PROVIDE_NFT_INFORMATION.
+
+        Parameters
+        ----------
+            -> Check documentation of APDU
+
+        Returns
+        -------
+        bytes
+            APDU command for PROVIDE_NFT_INFORMATION.
+
+        """
+
+        cdata: bytes = plugin.serialize()
+
+        return self.serialize(cla=self.CLA,
+                              ins=InsType.INS_PROVIDE_NFT_INFORMATION,
+                              p1=0x00,
+                              p2=0x00,
+                              cdata=cdata)
+
+    def provide_erc20_token_information(self, info: ERC20_Information):
+        """Command builder for PROVIDE_ERC20_INFORMATION.
+
+        Parameters
+        ----------
+            -> Check documentation of APDU
+
+        Returns
+        -------
+        bytes
+            APDU command for PROVIDE_ERC20_INFORMATION.
+
+        """
+
+        cdata: bytes = info.serialize()
+
+        return self.serialize(cla=self.CLA,
+                              ins=InsType.INS_PROVIDE_ERC20,
+                              p1=0x00,
+                              p2=0x00,
+                              cdata=cdata)
+
+    def get_public_key(self, bip32_path: str, display: bool = False) -> bytes:
+        """Command builder for GET_PUBLIC_KEY.
+
+        Parameters
+        ----------
+        bip32_path: str
+            String representation of BIP32 path.
+        display : bool
+            Whether you want to display the address on the device.
+
+        Returns
+        -------
+        bytes
+            APDU command for GET_PUBLIC_KEY.
+
+        """
+        bip32_paths: List[bytes] = bip32_path_from_string(bip32_path)
```

Comment:
It removes the principle that a function does only one thing at a time.
When I call `bip32_path_from_string` I want it to return the bip32 path and not the join of the bip32 path and its size

because we may want to call the function without the size later

Reply:
Just like early optimisation, writing code to fit possible usages instead of actual ones is a bad idea. Bip32 paths are **always** packed with their size in APDUs, as their length is variable.

But you're right about the meaning of the name of this function, which should return the bip32 path only (if I were a rigid soul I would even ask it to be be called `bip32_path_byte_array_from_string` as `bip32_path` alone does not say anything about its format).

If you want to keep this function for later use, just create a new function:
```python
def packed_bip32_path_from_string(path: str) -> bytes:
    bip32_path = bip32_path_from_string(path)
    return b"".join([
            to_bytes(len(bip32_paths), size=1),
            *bip32_paths
        ])
```

Reply:
https://github.com/LedgerHQ/app-ethereum/pull/275/commits/986b4b427ecb9acc838951ee7b633ff36e6fbdf8